### PR TITLE
[pipeline] update custom pipeline

### PR DESCRIPTION
**Pull request by star-e**
Added custom effect and native pipeline.
Refactored ProgramLib, moved functions to ProgramUtils.
Changed features related to descriptor set, generally it has no performance impact on current pipeline.
Adjusted NativePipeline folder structures.

### Changelog

* update native pipeline

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

**Comment by star-e** [comment]
@cocos-robot run test cases

**Comment by star-e** [comment]
@cocos-robot run test cases

**Comment by star-e** [comment]
@cocos-robot run test cases

**Comment by star-e** [comment]
@cocos-robot run test cases

**Comment by star-e** [comment]
@cocos-robot run test cases

**Comment by star-e** [comment]
@cocos-robot run test cases

**Comment by star-e** [comment]
@cocos-robot run test cases

**Comment by star-e** [comment]
@cocos-robot run test cases

**Comment by star-e** [comment]
@cocos-robot run test cases

**Comment by star-e** [comment]
@cocos-robot run test cases

**Comment by star-e** [comment]
@cocos-robot run test cases

File: cocos/asset/assets/effect-asset.ts [hunk 1]
```diff
@@ -302,6 +302,7 @@ export class EffectAsset extends Asset {
      */
     public onLoaded () {
         if (cclegacy.rendering && cclegacy.rendering.enableEffectImport) {
+            addEffectDefaultProperties(this);
```
Comment (star-e) [inline cocos/asset/assets/effect-asset.ts:305]: This change fixes missing logic compared with previous pipeline

File: cocos/render-scene/core/pass.ts [hunk 2]
```diff
@@ -124,11 +124,7 @@ export class Pass {
         if (info.stage !== undefined) { pass._stage = info.stage; }
         if (info.dynamicStates !== undefined) { pass._dynamicStates = info.dynamicStates; }
         if (info.phase !== undefined) {
-            if (cclegacy.rendering && cclegacy.rendering.enableEffectImport) {
-                pass._phase = cclegacy.rendering.completePhaseName(info.phase);
-            } else {
-                pass._phase = getPhaseID(info.phase);
-            }
+            pass._phase = getPhaseID(info.phase);
```
Comment (star-e) [inline cocos/render-scene/core/pass.ts:127]: Reverted previous change.

File: cocos/render-scene/core/pass.ts [hunk 3]
```diff
@@ -595,20 +591,25 @@ export class Pass {
         this._stage = RenderPassStage.DEFAULT;
         if (cclegacy.rendering && cclegacy.rendering.enableEffectImport) {
             const r = cclegacy.rendering;
-            this._phase = r.completePhaseName(info.phase);
-            this._passID = r.getPassID(info.pass);
+            if (typeof info.phase === 'number') {
+                this._passID = (info as Pass)._passID;
+                this._phaseID = (info as Pass)._phaseID;
```
Comment (star-e) [inline cocos/render-scene/core/pass.ts:596]: if info.phase is number, the info must be type of `Pass`. This is determined by the code.

File: cocos/render-scene/core/pass.ts [hunk 4]
```diff
@@ -595,20 +591,25 @@ export class Pass {
...
+            } else {
+                this._passID = r.getPassID(info.pass);
```
Comment (star-e) [inline cocos/render-scene/core/pass.ts:598]: info is not `Pass`, we need to initialize `passID` and `phaseID`

File: cocos/render-scene/core/pass.ts [hunk 5]
```diff
@@ -838,6 +839,8 @@ export class Pass {
     get primitive (): PrimitiveMode { return this._primitive; }
     get stage (): RenderPassStage { return this._stage; }
     get phase (): number { return this._phase; }
+    get passID (): number { return this._passID; }
+    get phaseID (): number { return this._phaseID; }
```
Comment (star-e) [inline cocos/render-scene/core/pass.ts:843]: These are only used in custom pipeline.

File: cocos/render-scene/core/program-utils.ts [hunk 6]
```diff
@@ -114,8 +115,28 @@ export function getVariantKey (programInfo: IProgramInfo, defines: MacroRecord)
...
+            return prevSize + GetTypeSize(m.type) * count;
+        }
+        console.error(`uniform '${m.name}' must have a count`);
+    }
+    return prevSize;
+}
+
 export function getSize (blockMembers: Uniform[]) {
-    return blockMembers.reduce((s, m) => s + GetTypeSize(m.type) * m.count, 0);
+    return blockMembers.reduce(getUniformSize, 0);
```
Comment (star-e) [inline cocos/render-scene/core/program-utils.ts:139]: UniformSize might change according to hardware. We need to get size info at runtime.

File: cocos/rendering/define.ts [hunk 7]
```diff
@@ -755,4 +755,8 @@ export function supportsR32FloatTexture (device: Device) {
         && !(device.gfxAPI === API.WEBGL); // wegl 1  Single-channel float type is not supported under webgl1, so it is excluded
 }
 
+export function isEnableEffect (): boolean {
+    return !!(cclegacy.rendering && cclegacy.rendering.enableEffectImport);
+}
```
Comment (star-e) [inline cocos/rendering/define.ts:760]: help function to check custom effected enabled. Will be removed once custom pipeline is finished.

File: native/cocos/renderer/core/ProgramLib.cpp [hunk 8]
```diff
@@ -148,29 +145,6 @@ int32_t getSize(const IBlockInfo &block) {
...
-    // samplerTexture handles
-    for (const auto &samplerTexture : tmpl.samplerTextures) {
-        handleMap[samplerTexture.name] = genHandle(samplerTexture.binding,
-                                                   samplerTexture.type,
-                                                   samplerTexture.count);
-    }
-    return handleMap;
-}
-
 bool dependencyCheck(const ccstd::vector<ccstd::string> &dependencies, const MacroRecord &defines) {
```
Comment (star-e) [inline native/cocos/renderer/core/ProgramLib.cpp:148]: moved to file "ProgramUtils.h"

File: native/cocos/core/assets/EffectAsset.h [hunk 9]
```diff
@@ -535,7 +536,7 @@ struct IShaderInfo {
     ccstd::vector<IBufferInfo> buffers;
     ccstd::vector<IImageInfo> images;
     ccstd::vector<IInputAttachmentInfo> subpassInputs;
-    // ccstd::vector<IDescriptorInfo> descriptors;
+    ccstd::vector<IDescriptorInfo> descriptors;
```
Comment (star-e) [inline native/cocos/core/assets/EffectAsset.h:539]: fixed jsb crash caused by effect compiler. descriptors now can be added back.

**Comment by star-e** [comment]
@cocos-robot run test cases

File: cocos/render-scene/core/pass.ts (hunk 3 above)
Comment (minggo) [reply cocos/render-scene/core/pass.ts:596]: These codes only exists in ts?

Comment (star-e) [reply cocos/render-scene/core/pass.ts:596]: They will be added to native pipeline later.

Comment (star-e) [reply cocos/render-scene/core/pass.ts:596]: I haven't seen c++ code yet. I will properly handle these cast. Avoid them at best.

**Review by minggo — APPROVED** [review]

**Pull request merged**